types/decimal.go

```diff
@@ -0,0 +1,422 @@
+func (d Dec) Neg() Dec          { return Dec{new(big.Int).Neg(d.Int)} } // reverse the decimal sign
+
+// addition
+func (d Dec) Add(d2 Dec) Dec {
+	res := new(big.Int).Add(d.Int, d2.Int)
+
+	if res.BitLen() > 255+DecimalPrecisionBytes {
+		panic("Int overflow")
+	}
+	return Dec{res}
+}
+
+// subtraction
+func (d Dec) Sub(d2 Dec) Dec {
+	res := new(big.Int).Sub(d.Int, d2.Int)
+
+	if res.BitLen() > 255+DecimalPrecisionBytes {
+		panic("Int overflow")
+	}
+	return Dec{res}
+}
+
+// multiplication
+func (d Dec) Mul(d2 Dec) Dec {
+	mul := new(big.Int).Mul(d.Int, d2.Int)
+	chopped := chopPrecisionAndRound(mul)
+
+	if chopped.BitLen() > 255+DecimalPrecisionBytes {
+		panic("Int overflow")
+	}
+	return Dec{chopped}
+}
+
+// quotient
+func (d Dec) Quo(d2 Dec) Dec {
+
+	// multiply precision twice
+	mul := new(big.Int).Mul(d.Int, precisionReuse)
+	mul.Mul(mul, precisionReuse)
+
+	quo := new(big.Int).Quo(mul, d2.Int)
+	chopped := chopPrecisionAndRound(quo)
+
+	if chopped.BitLen() > 255+DecimalPrecisionBytes {
+		panic("Int overflow")
+	}
+	return Dec{chopped}
+}
+
+func (d Dec) String() string {
+	str := d.ToLeftPaddedWithDecimals(Precision)
+	placement := len(str) - Precision
+	if placement < 0 {
+		panic("too few decimal digits")
+	}
+	return str[:placement] + "." + str[placement:]
+}
+
+// TODO panic if negative or if totalDigits < len(initStr)???
+// evaluate as an integer and return left padded string
+func (d Dec) ToLeftPaddedWithDecimals(totalDigits int8) string {
+	intStr := d.Int.String()
+	fcode := `%0` + strconv.Itoa(int(totalDigits)) + `s`
+	return fmt.Sprintf(fcode, intStr)
+}
+
+// TODO panic if negative or if totalDigits < len(initStr)???
+// evaluate as an integer and return left padded string
+func (d Dec) ToLeftPadded(totalDigits int8) string {
+	chopped := chopPrecisionAndRound(d.Int)
+	intStr := chopped.String()
+	fcode := `%0` + strconv.Itoa(int(totalDigits)) + `s`
+	return fmt.Sprintf(fcode, intStr)
+}
+
+//     ____
+//  __|    |__   "chop 'em
+//       ` \     round!"
+// ___||  ~  _     -bankers
+// |         |      __
+// |       | |   __|__|__
+// |_____:  /   | $$$    |
+//              |________|
+
+// nolint - go-cyclo
+// Remove a Precision amount of rightmost digits and perform bankers rounding
+// on the remainder (gaussian rounding) on the digits which have been removed.
+//
+// TODO We should make this function  mutate the input. The functions here
+// don't need to allocate different memory for chopped after computing the
+// result
+func chopPrecisionAndRound(d *big.Int) *big.Int {
+
+	// remove the negative and add it back when returning
+	if d.Sign() == -1 {
+		// make d positive, compute chopped value, and then un-mutate d
+		d = d.Neg(d)
+		d = chopPrecisionAndRound(d)
+		d = d.Neg(d)
+		return d
+	}
+
+	// get the trucated quotient and remainder
+	quo, rem := big.NewInt(0), big.NewInt(0)
+	quo, rem = quo.QuoRem(d, precisionReuse, rem)
```

Comment:
We actually don't need to use `quo` here, it can just be `d`, and then `quo` never had to be allocated :)

Reply:
cool I'm going to make `quo, rem := d, big.NewInt(0)` just to keep it explicit - makes the return statements more clear too.

Reply:
okay - turns out this causes a few undesired side effects because often this function is expected to return a new decimal not modify the original - I'm going to leave as is for now, I feel like this is a pretty minor optimization - but feel free to make the modification and add the fixes (a bunch of lcd stuff fails)

Reply:
Sure! I'll look into this more once we get get the main PR merged!